store/store_action.go

```diff
@@ -385,30 +403,88 @@ func (s *Store) snapAction(ctx context.Context, currentSnaps []*CurrentSnap, act
 
 		aJSON.InstanceKey = instanceKey
 
-		actionJSONs[i] = aJSON
+		actionJSONs[actionIndex] = aJSON
+		actionIndex++
 	}
 
+	groupingsAssertions := make(map[string]*snapActionJSON)
+
 	// assertions
 	var assertMaxFormats map[string]int
 	if len(toResolve) > 0 {
-		i := len(actionJSONs) - len(toResolve)
 		for grp, ats := range toResolve {
 			aJSON := &snapActionJSON{
 				Action: "fetch-assertions",
 				Key:    string(grp),
 			}
-			aJSON.Assertions = make([]assertAtJSON, len(ats))
+			aJSON.Assertions = make([]interface{}, len(ats))
+			groupingsAssertions[aJSON.Key] = aJSON
+
 			for j, at := range ats {
-				aJSON.Assertions[j].Type = at.Type.Name
-				aJSON.Assertions[j].PrimaryKey = at.PrimaryKey
+				aj := &assertAtJSON{
+					Type:       at.Type.Name,
+					PrimaryKey: at.PrimaryKey,
+				}
 				rev := at.Revision
 				if rev != asserts.RevisionNotKnown {
-					aJSON.Assertions[j].IfNewerThan = &rev
+					aj.IfNewerThan = &rev
 				}
+				aJSON.Assertions[j] = aj
 			}
-			actionJSONs[i] = aJSON
-			i++
+			actionJSONs[actionIndex] = aJSON
+			actionIndex++
 		}
+	}
+
+	if len(toResolveSeq) > 0 {
+		for grp, ats := range toResolveSeq {
+			key := string(grp)
+			// append to existing grouping if applicable
+			aJSON := groupingsAssertions[key]
+			existingGroup := aJSON != nil
+			if !existingGroup {
+				aJSON = &snapActionJSON{
+					Action: "fetch-assertions",
+					Key:    key,
+				}
+				aJSON.Assertions = make([]interface{}, 0, len(ats))
+				actionJSONs = append(actionJSONs, aJSON)
+			}
+			for _, at := range ats {
+				aj := assertSeqAtJSON{
+					Type:        at.Type.Name,
+					SequenceKey: at.SequenceKey,
```

Comment:
Do we need to check whether len(at.SequenceKey) > 0 and return an internal error if it's not?

Reply:
I don't think so, at this point we trust both sequence keys and primary keys (compare with the existing code above) that we receive from the pool (pool would error out much earlier if they were wrong).